### PR TITLE
add NWBHDF5IO.read_nwb() method

### PR DESCRIPTION
**Pull request by h-mayorquin**
This was discussed in #1974

This is the first part of the program described by @oruebel [here](https://github.com/NeurodataWithoutBorders/pynwb/issues/1974#issuecomment-2455440137). In a future PR and once https://github.com/hdmf-dev/hdmf-zarr/issues/225 is merged I will implement `pynwb.read_nwb` in another PR.

Meanwhile, I have two questions:
1) What streaming method should be used when an s3 path is passed?
2) For testing this, the [ros3 test](https://github.com/h-mayorquin/pynwb/blob/7c65fe2cb20f10cefa0cefeae019ce32c6b1706e/tests/integration/ros3/test_ros3.py#L10-L20) have some files that I think I could use, is that OK? 

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [x] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR clearly describes the problem and the solution?
- [x] Is your contribution compliant with our coding style? This can be checked running `ruff check . && codespell` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?
- [x] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

### CODE REVIEW

**Comment by oruebel** [comment]
> 2\. For testing this, the [ros3 test](https://github.com/h-mayorquin/pynwb/blob/7c65fe2cb20f10cefa0cefeae019ce32c6b1706e/tests/integration/ros3/test_ros3.py#L10-L20) have some files that I think I could use, is that OK?

Yes, that should be fine

> 1. What streaming method should be used when an s3 path is passed

I think we should probably use `fsspec` as the default https://pynwb.readthedocs.io/en/latest/tutorials/advanced_io/streaming.html#streaming-method-1-fsspec 

If installed, we could also use `remfile` https://pynwb.readthedocs.io/en/latest/tutorials/advanced_io/streaming.html#method-3-remfile  and use `fsspec` as the fall-back. `remfile` tends to be a bit faster than `fsspec`, but I think `fsspec` is more widely used. 

@bendichter what would you suggest for this?

**Comment by oruebel** [comment]
Aside from adding s3 and testing, the outline you started here looks good to me.

**Comment by h-mayorquin** [comment]
OK, I added s3 testing.

**Comment by h-mayorquin** [comment]
Is the test failing a permission issue?
image

**Comment by rly** [comment]
Seems like it was a codecov issue with detecting forks, perhaps when commits happen in quick succession: https://github.com/codecov/codecov-action/issues/1580 . I re-ran the failing runs.

File: tests/integration/hdf5/test_io.py
Comment (rly) [inline tests/integration/hdf5/test_io.py]:
```suggestion
    def test_read_nwb_method_s3_path(self):
```

File: requirements-opt.txt
Comment (oruebel) [inline requirements-opt.txt:8]: Are ``requests`` and ``aiohttp`` needed here or these being installed by fsspec? I don't see these libraries being used directly here and if ``fsspec`` installs them then I don't think we should need to specify these separately.

Comment (rly) [reply requirements-opt.txt:8]: fsspec does not install them. they are required for http streaming

File: tests/integration/hdf5/test_io.py
Comment (oruebel) [inline tests/integration/hdf5/test_io.py:624]: @rly does Codecove run this test when `fsspec` is specified as an optional requirement or would we need to add `fsspec` elsewhere too? I'm just wondering because codecov seems to report the s3 parts of the code changes as not covered above, even though this test should cover them.

**Review by oruebel — APPROVED** [review]
Aside from the codecov question, this looks good to me. I approve the PR, because I assume the codecov question can probably be resolved in the comment, rather than requiring additional code changes, and it seems easier for me to re-approve if there are changes, rather than holding up the PR.

File: tests/integration/hdf5/test_io.py
Comment (rly) [reply tests/integration/hdf5/test_io.py:624]: It does run this. the codecov "Files Changed" annotation seems to be inconsistent with their website: https://app.codecov.io/gh/NeurodataWithoutBorders/pynwb/pull/1979?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=NeurodataWithoutBorders

Comment (oruebel) [reply tests/integration/hdf5/test_io.py:624]: Thanks @rly for the clarification.

**Pull request merged**
